File: SWEA/D4/1218. ［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기/［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기.py
t = 10
for tc in range(1,t+1):
    n = int(input())
    lst = list(input())
    answer = 1
    stack = []
    for l in lst:
        if l in ['(','{','[','<']:
            stack.append(l)
        elif l == ')' and stack[-1] == '(':
            stack.pop()
        elif l == '}' and stack[-1] == '{':
            stack.pop()
        elif l == ']' and stack[-1] == '[':
            stack.pop()
        elif l == '>' and stack[-1] == '<':
            stack.pop()
        else:
            answer = 0
            break
    
    print(f'#{tc} {answer}')

File: SWEA/D4/1218. ［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기/［S／W 문제해결 기본］ 4일차 － 괄호 짝짓기.java

import java.io.*;
import java.util.*;

public class Solution {

	public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int t = 10;

        for (int tc = 1; tc <= t; tc++) {
            int n = Integer.parseInt(br.readLine());
            String lst = br.readLine();
            int answer = 1;
            Deque<Character> deq = new ArrayDeque<>();

            for (char l : lst.toCharArray()) {
                if (l == '(' || l == '{' || l == '[' || l == '<') {
                	deq.addLast(l);
                } else if (l == ')' && deq.peekLast() == '(') {
                	deq.pollLast();
                } else if (l == '}' && deq.peekLast() == '{') {
                	deq.pollLast();
                } else if (l == ']' && deq.peekLast() == '[') {
                	deq.pollLast();
                } else if (l == '>' && deq.peekLast() == '<') {
                	deq.pollLast();
                } else {
                    answer = 0;
                    break;
                }
            }

            System.out.println("#" + tc + " " + answer);
        }
    }
}
